Add file magic and support for configurable compression algorithms

fako1024 opened the pull request:
Note: breaking change, only works on a newly created database for now (conversion tool TBD). Would be great if you could do a few tests on a fresh DB on your side and do a quick review of the changes. Will look into conversion tool asap...

fako1024 commented:
Come to think of it, ignore that. The more I look at this the less I like it. I'll revamp the whole GPFile structure, now that we need a conversion tool I can improve a lot here... Give me a week or so...

els0r commented:
fko is on a roll!

On first glance, tests with the new format checked out (tested on Debian 10 so far).

I was quite surprised how "little" changes were necessary to get this in. Looks like we were closer than we thought.

Let me know if I can contribute for the conversion utility or if you need to discuss anything regarding .gpf v2.

And thanks a lot for tackling this. Cool to have some movement on this again 😁.

els0r reviewed (approved):
Good to go from my end.

Of course interested in further tweaks of .gpf v2.

fako1024 commented:
@els0r Alright, here we go. New format ready to use, tests fixed, converter provided (cmd/legacy). As simple as:
`./legacy -path /path/to/godb`
for a dry run (Will do all the conversion steps save altering any files), and when you're sure your're ready:
`./legacy -path /path/to/godb -dry-run=false`
Tested on the testdb in the repo, a full backup of the whole godb prior to conversion is _highly_ recommended (and goProbe should not be running to avoid any race conditions). After conversion, new version of goProbe + goQuery should work as before.

els0r commented:
Awesome work on making this more useful and leaner.

The conversion tool currently fails on a large (!) DB from my NAS. After running for a few minutes, it craps out with the following error:

```
$ legacy -path /tmp/db -dry-run=false     
2020/08/25 20:01:16 File /tmp/db/docker0/1554595200/bytes_rcvd.gpf already converted, skipping...
2020/08/25 20:01:16 Error converting legacy file /tmp/db/docker0/1554595200/bytes_sent.gpf: rename /tmp/db/docker0/1554595200/bytes_sent.gpf.tmp /tmp/db/docker0/1554595200/bytes_sent.gpf: no such file or directory
```

What is even more curious is what happens when I execute the same command on the same DB. It will cycle through the error, first with bytes_sent, then bytes_rcvd, then sip, then dip, etc.

When I check, the .gpf file does exist and is converted and the .meta file is written.

fako1024 commented:
Thanks for testing, man! I'll try to reproduce the issue on my, also large (!) DB. I already have a suspicion though (file system race condition, seen something similar in a different situation with Go :dizzy_face:). I assume that if you retry the same (clean, pre-conversion) source DB, it doesn't happen at all or at least on a different file. Would be cool if you could try that to confirm, just in case I can't reproduce the issue...

fako1024 commented:
Problem was found, quite easy actually - those files did not contain any flow data (i.e. there was no traffic measured in the covered period of time, just the timestamps were added to the file header). In the new gpfile concept, this leads to a missing data file (only header file exists), hence the Rename() fails (and the file you are seeing is actually the old, unconverted one). Implemented a fix for the conversion tool (and also improved write performance for these bulk operations a bit) and running tests right now, will push asap.

fako1024 commented:
@els0r Done, should work like a charm now (and even handle corrupt files gracefully, throwing a warning)...

fako1024 commented:
Urgh, that adds quite some size overhead due to the meta files. I'll do one more little change, don't final-convert your DB just yet :-P

fako1024 commented:
Alright, now I'm done for the moment. Reduce the DB size pre-conversion 8.0 GB to 6.8 GB post-conversion (due to smaller overhead after optimizations and several empty blocks in my DB (some low-traffic interfaces)). Feel free to give it a shot!

els0r commented:
Conversion worked flawlessly now. Thanks for the fix.

The overhead of the `.meta` files was not noticeable in my case. Still useful to have the JSON in the shortened format. After all, those files were always supposed to be mainly machine readable. I also briefly had a look at BSON, but it is not a given that this reduces storage overhead of the meta files.

I'll deploy the version from the latest commit on my infrastructure today and monitor. Will let you know if anything pops up.

Can't wait to merge this and start supporting more compression algorithms 🤤. 

Kudos!

florianl commented on cmd/legacy/main.go:
```diff
@@ -82,6 +82,9 @@ func convert(path string, dryRun bool) error {
 	defer legacyFile.Close()
 
 	tmpPath := path + ".tmp"
+	os.Remove(tmpPath)
```
maybe add a `defer` before `os.Remove()`. Also for the next line.

fako1024 commented:
Cool! Maybe you could do a few random comparisons for some fixed time ranges (long ones) using old goQuery on the old DB and new goQuery on the new DB just to get some more confidence that all is cool? I'll do the same on my end. Regarding the format: Maximum size and performance optimization at the same time could be reached by using a custom text format, e.g.
```
<version>
<p>
<ts1>,<p>,<l>,<r>[,<c>]
...
```
This removes all overhead caused by JSON etc., can be read line by line (i.e. doesn't require reflect like json(iter)) and hence is faster, while at the same time retaining full human readability. I'd be open to implement that, just wasn't sure if that'd hit you approval because it's non-standard :-P

Your call...

els0r commented:
Yo. Fine by me. It is helpful to have a format that is semi-readable (e.g. doesn't need hexdump), which your proposal would totally tackle.

Also, the parsing would have to be implemented only once and can be reused across the code base. Does sound like best of both worlds to me. I'd say `go` for it 😉.

On another note, I'm still running into errors when converting (segfaults from lz4). I'll look into if its a corrupted block and never worked in the first place (with old `goQuery`) or whether this may be a shortcoming of the conversion tool.

As for finishing up this baby, I still see the following steps:

- @fako1024: Tweak the header files 
- @els0r:  More testing with conversion and reading out (investigate the errors I'm seeing)
- @els0r: describe new format and conversion steps into `README.md`. The main description can be part of the release notes of `v4.0.0`

Deal?

fako1024 commented:
@els0r Alright, will do and update the PR. Regarding the segfaults - I had that once or twice, too... In my case it was a corrupt file, combined with crappy handling in C land. If you just continue the conversion (it is resumable after interruption and will just skip everything already converted) it worked fine though for me. Since I wasn't able to reproduce it reliably I couldn't investigate yet...

fako1024 commented:
@els0r Alrighty, all done for now. On top of the serialization format (brings down my DB from 8GB to 6.4GB due to the reduction in size and the large number of empty blocks and increases round-trip read/write by about a factor of 2-3) I've fixed the LZ4 crashes by updating to the newest version, improved writeout performance by using buffering and added a little gimmick: ZSTD compressor, ready to use, this time no hacky stuff, but the standard C wrapper package that's actively maintained. Enjoy (and test thoroughly, all of it :-P).

els0r commented:
Good stuff. Deployed on my NAS for now and testing with ZSTD:

- initial tests look good (DB conversion works, querying of flows compressed with ZSTD works)
- DB conversion was __fast__ this time around. It never ceases to amaze me how much a small go routine/worker setup can improve performance. Even when I/O intensive workloads are in the mix

As for the use of `bufio`, I'm really curious how that improved overall performance/memory footprint. We can chat about that offline.

I'll make sure that `lz4_darwing.a` makes it into the repo and will test once I have access to my Mac (coding on iPad atm.).

After running final tests on router, you have my blessings to merge 😃.

Thanks again. This is really cool!

els0r reviewed (approved):
I'll add `lz4_darwin.a` and we are ready to go.

els0r commented:
Tests on Mac OS X also completed (both ZSTD and LZ4).
